### PR TITLE
fix:  transfer amount less than or equal to the balance.

### PR DESCRIPTION
**Pull request by riflecode**
### What changed? Why?
The user should be able to transfer the entire balance from their account.

image

### CODE REVIEW

**Comment by cb-heimdall** [comment]
<h1>🟡 Heimdall Review Status</h1>
<table>
<tr>
<th>Requirement</th>
<th>Status</th>
<th>More Info</th>
</tr>
<tr>
<td>Reviews</td>
<td>
🟡
<code>0/1</code>
</td>
<td>
<details>
<summary>Denominator calculation</summary>
<table>
<tr>
<td>
<details>
<summary>Show calculation</summary>
<table>
<tr>
<td><code>1</code> if user is bot</td>
<td>0</td>
</tr>
<tr>
<td><code>1</code> if user is external</td>
<td>0</td>
</tr>
<tr>
<td>
From <code>.codeflow.yml</code>
</td>
<td>1</td>
</tr>
<tr>
<td>
Additional review requirements
<details>
<summary>Show calculation</summary>
<table>
<tr>
<td><strong>Max</strong></td>
<td><strong>0</strong></td>
</tr>
</table>
</details>
</td>
<td>0</td>
</tr>
<tr>
<td>
From <code>CODEOWNERS</code>
</td>
<td>0</td>
</tr>
<tr>
<td>Global minimum</td>
<td>0</td>
</tr>
<tr>
<td><strong>Max</strong></td>
<td>
<strong>
1
</strong>
</td>
</tr>
</table>
</details>
</td>
<td>1</td>
</tr>
<tr>
<td><code>1</code> if commit is unverified</td>
<td>0</td>
</tr>
<tr>
<td><strong>Sum</strong></td>
<td><strong>1</strong></td>
</tr>
</table>
</details>
</td>
</tr>
</table>

**Comment by John-peterson-coinbase** [comment]
Thank you for your contribution @riflecode . Can you please add a test to `tests/test_wallet_address.py` for ensure balance on the full amount? Thanks!

Something like:
```
@patch("cdp.Cdp.api_clients")
def test_ensure_sufficient_balance_sufficient_full_amount(
    mock_api_clients, wallet_address_factory, balance_model_factory
):
    """Test the ensure_sufficient_balance method with sufficient full amount balance."""
    wallet_address = wallet_address_factory()
    balance_model = balance_model_factory(
        amount="1500000000000000000", network_id="base-sepolia", asset_id="eth", decimals=18
    )

    mock_get_balance = Mock()
    mock_get_balance.return_value = balance_model
    mock_api_clients.external_addresses.get_external_address_balance = mock_get_balance

    wallet_address._ensure_sufficient_balance(Decimal("1.5"), "eth")

    mock_get_balance.assert_called_once_with(
        network_id=wallet_address.network_id, address_id=wallet_address.address_id, asset_id="eth"
    )
```

**Comment by riflecode** [comment]
> Thank you for your contribution @riflecode . Can you please add a test to `tests/test_wallet_address.py` for ensure balance on the full amount? Thanks!
> 
> Something like:
> 
> ```
> @patch("cdp.Cdp.api_clients")
> def test_ensure_sufficient_balance_sufficient_full_amount(
>     mock_api_clients, wallet_address_factory, balance_model_factory
> ):
>     """Test the ensure_sufficient_balance method with sufficient full amount balance."""
>     wallet_address = wallet_address_factory()
>     balance_model = balance_model_factory(
>         amount="1500000000000000000", network_id="base-sepolia", asset_id="eth", decimals=18
>     )
> 
>     mock_get_balance = Mock()
>     mock_get_balance.return_value = balance_model
>     mock_api_clients.external_addresses.get_external_address_balance = mock_get_balance
> 
>     wallet_address._ensure_sufficient_balance(Decimal("1.5"), "eth")
> 
>     mock_get_balance.assert_called_once_with(
>         network_id=wallet_address.network_id, address_id=wallet_address.address_id, asset_id="eth"
>     )
> ```

Thank you. I will improve this test case

**Review by John-peterson-coinbase — APPROVED** [review]

File: tests/test_wallet_address.py [hunk 1]
```diff
@@ -999,3 +999,23 @@ def test_deploy_multi_token_broadcast_api_error(mock_smart_contract, wallet_addr
     mock_smart_contract.create.assert_called_once()
     mock_smart_contract_instance.sign.assert_called_once_with(wallet_address_with_key.key)
     mock_smart_contract_instance.broadcast.assert_called_once()
+
```
Comment (John-peterson-coinbase) [inline tests/test_wallet_address.py:1002]:
```suggestion


```

Ruff formatting nit

**Comment by John-peterson-coinbase** [comment]
@riflecode - The CodeQL CI check did not get triggered since this is your first contribution to the repo. Can you push up a fix to https://github.com/coinbase/cdp-sdk-python/pull/36#discussion_r1820941978 formatting issue and that should trigger the CodeQL CI run so that we can merge. 

Thanks!

**Review by John-peterson-coinbase — APPROVED** [review]

**Pull request merged**